### PR TITLE
Eth2-to-Near-relay: prototype implementation

### PR DESCRIPTION
**Pull request by olga24912**
# Eth2 to Near Relay prototype

## Changes
* Implemented `eth2-to-near-relay`;
* `eth-types`: remove redundant feature flags on `serde::Serialize` implementation

## General description

In this PR, a simple eth2 to near relay is implemented. This relay transfers from Eth to Near: (1) eth1 block header each slot, (2) finality light client updates each epoch, (3) sync_commite_update each period. 

## Relay CLI
**File:** _main.rs_
The point of entry for working with the relay. To get more information about the command line interface, you can run `cargo run -- --help`. For parsing command line arguments, we use `clap` crate.  

## Main logic
**File:** _eth2near_relay.rs_
Here you can find the main logic for the relay. Struct Eth2NearRelay contains two public methods: (1) _init_ for `Eth2NearRelay` construction, and (2) _run_ for execution. By design, the _run_ method is infinity and takes block headers, and light client updates from Ethereum Block Chain and sends it to the Ethereum Light Client on NEAR. At each loop iteration, it (1) searches for the last known header on the ETH client on NEAR, (2) fetch and send a bunch of eth1 block headers, and (3) send, if necessary new light client update. 

## Eth on near client contract wrapper
**File:** _eth_client_contract.rs_, _init_contract.rs_
In the struct `EthClientContract`, we encapsulate interaction with Ethereum Light Client in NEAR. Maybe in the future makes sense to create `EthClientContract` trait and next implementations (1) `DirectEthClientContract`, (2) `DAOEthClientContract`, (3) `FileEthClientContract`. For now, it directly interacts with [this version](https://github.com/aurora-is-near/rainbow-bridge-private/blob/effacc5702ddb2a8d8df9c76f121b133c4b76d26/contracts/near/eth2-client/src/lib.rs) of the contract. 

Public API of `EthClientContract`: 
* _is_known_block_ -- check if this eth1 execution block hash is in the list of unfinalized blocks in the contract 
* _send_light_client_update_ -- for sending light client update
* _get_finalized_beacon_block_hash_ -- for getting the beacon block hash of the last known finalized block on the contract
* _send_headers_ -- for sending the bunch of eth1_block_headers
* _register_ -- for registration of the current relay at the contract
* _init_contract_ -- for initializing the contract

For interaction with contract, we use _near_jsonrpc_client_ crate. We have two methods for direct interaction with the contract: (1) _call_change_method_ -- for sending a transaction to the change method on the contract, and (2) _call_change_method_ -- for calling a view method.  

## Eth RPC API
**Files:** _beacon_rpc_client.rs_, _eth1_rpc_client.rs_
### Eth1 API && Eth2 API
For getting Eth1 execution block headers, we should use eth1 API, and for getting Light Client Updates, we need beacon API. And such API is provided by different endpoints. So we need two different endpoints and two different clients: 
* beaconRPCClient -- for beacon API
* Eth1RPCClient -- for eth1 API

### Light client update API
BeaconRPCClient implemented a series of methods that gets LightClientUpdate from Light Client Server (full node with LightClient API).

For using it, the endpoint should provide Light Client API. Not all Beacon endpoints have such an API. Moreover, there can be many ambiguities with API. We use the API from lodestar (can be found [here](https://github.com/ChainSafe/lodestar/blob/bdfe5519182ef6ab6786141267d540e281925906/packages/api/src/beacon/routes/lightclient.ts#L62) ). 

For getting Light Client Update, we have next methods:
* _get_light_client_update_  -- this method returns LightClientUpdate for period. But it contains the best finality header, NOT THE LAST one. 
* _get_finality_light_client_update_ -- this method returns LightClientUpdate with the last finality header but without sync commity. 
* _get_finality_light_client_update_with_sync_commity_update_ -- it takes LightClientUpdate from _get_finality_light_client_update_ and the sync commity from _get_light_client_update_

Current lodestar Light Client API is a little outdated and doesn't provide [signature_slot](https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/sync-protocol.md). For getting `signature_slot` we brute force slots and compare SyncAggregation in BeconBlockBody and in LightClientUpdate. 

In tests in `BeaconRPCClient`, you can find two ignored tests we use for printing LightClientUpdate JSON and JSON for Headers with Eth1Data in some range. 

## LightClientUpdate -- Json Serialization
**Files:** _contracts/near/eth-types/eth2.rs_, _contracts/near/eth-types/macros.rs_

To be able to save LightClientUpdate to a JSON file(for example, for testing) we also add _serde_ JSON serialization for arrays similar to deserialization.

## Execution block proof
**Files:**  _beacon_block_body_merkle_tree.rs_, _beacon_block_header_with_execution_data.rs_, _execution_block_proof.rs_
IMG_20220712_170534

When we send light client update for finality block inside light client update, we also send Eth1 execution block hash with the Merkle proof of include to Beacon Block Body. Execution block hash you can find at _BeaconBlockBody.execution_payload.block_hash._ 

So, for creating Merkle proof, we need two levels of Merkle Tree, as shown in the picture. Both Merkle trees you can find in _beacon_block_body_merkle_tree.rs_  The first level Merkle tree for beacon block body and the second level Merkle tree for execution payload. 

The execution block hash proof creation you can find in _execution_block_proof.rs_ First, we build two Merkle trees and concatenate together the Merkle proof for block_hash in execution_payload and the Merkle proof of execution_payload in beacon_block_body. The final Merkle proof is shown by the orange vertices on the picture; the orange numbers in the picture are the order of hashes in the proof.  

_beacon_block_header_with_execution_data.rs_ contain a structure which consists of beacon_block_header and correspondent execution_block_hash with Merkle proof. This structure is created for finality blocks in a light client update. 

## Logging
**Files:** _logger.rs_

For logging, we use the _log_ crate and just copy SimpleLogger with a small modification from this [doc](https://docs.rs/log/latest/log/#implementing-a-logger). 

Currently, it just prints all logs in the console, but in the future, it can be improved to write it in the file as well. 

We use two types of logging: _info_(for general message) and _warn_(if some error happens). 

Also, In all `info!` and `warn!` we add `target: "relay"` for separating my logs message from messages inside libraries that we use. 

For handling errors, we have two strategies: 
(1) just unwrap the error at the beginning if the error occurs before we actually start the relay(during initialization, for example).  In that case, we can just read the backtrace and restart the relay. 
(2) write to log and try again for errors that happen during relays works.

### CODE REVIEW

File: eth2near/contract_wrapper/src/near_contract_wrapper.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        for i in 0..method_name.len() {
+            actions.push(
+                Action::FunctionCall(FunctionCallAction{
+                    method_name: method_name[i].clone(),
+                    args: args[i].clone(),
+                    gas: 300_000_000_000_000/method_name.len() as u64, // 300 TeraGas
+                    deposit: deposit[i].clone(),
+                })
+            );
+        }
```
Comment (sept-en) [inline eth2near/contract_wrapper/src/near_contract_wrapper.rs:94]: Readability needs to be improved

```suggestion

        let num_blocks_in_batch = method_name.len() as u64;
        let attached_gas_per_promise_in_batch = 300_000_000_000_000 / num_blocks_in_batch;
        let mut actions = Vec::new();
        for i in 0..method_name.len() {
            actions.push(
                Action::FunctionCall(FunctionCallAction{
                    method_name: method_name[i].clone(),
                    args: args[i].clone(),
                    gas: attached_gas_per_promise_in_batch,
                    deposit: deposit[i].clone(),
                })
            );
        }
```

File: eth2near/eth2near-block-relay-rs/olga_local_test_config.toml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+beacon_endpoint = "https://lodestar-kiln.chainsafe.io"
+eth1_endpoint = "https://rpc.kiln.themerge.dev"
+total_submit_headers = 4
+near_endpoint = "https://rpc.testnet.near.org"
+signer_account_id = "olga24912.testnet"
+path_to_signer_secret_key = "/home/olga/.near-credentials/testnet/olga24912.testnet.json"
+contract_account_id = "dev-1659151769901-77594397083671"
+network = "kiln"
+contract_type = "near"
+light_client_updates_submission_frequency_in_epochs = 2
```
Comment (sept-en) [inline eth2near/eth2near-block-relay-rs/olga_local_test_config.toml:10]: This file is redundant. If it's needed, it should be called `local_test_config.toml` and then should be in `.gitignore`.

File: eth2near/eth2near-block-relay-rs/src/beacon_rpc_client.rs [hunk 3]
```diff
@@ -1,27 +1,94 @@
...
 pub struct BeaconRPCClient {
-    endpoint_url: std::string::String,
+    endpoint_url: String,
+    client: Client,
 }
 
 impl BeaconRPCClient {
     const URL_HEADER_PATH: &'static str = "eth/v1/beacon/headers";
     const URL_BODY_PATH: &'static str = "eth/v2/beacon/blocks";
+    const GET_LIGHT_CLIENT_UPDATE_API: &'static str = "eth/v1/light_client/updates";
```
Comment (sept-en) [inline eth2near/eth2near-block-relay-rs/src/beacon_rpc_client.rs:80]: There should be consistency in the naming.

```suggestion
    const URL_GET_LIGHT_CLIENT_UPDATE_API: &'static str = "eth/v1/light_client/updates";
```

File: eth2near/eth2near-block-relay-rs/src/beacon_rpc_client.rs [hunk 4]
```diff
@@ -49,21 +116,119 @@ impl BeaconRPCClient {
...
     }
 
+    /// Returns `LightClientUpdate` struct for the given `period`.
+    ///
+    /// # Arguments
+    ///
+    /// * `period` - period id for which LightClientUpdate is fetched.
+    /// In Mainnet one period consist of 256 epochs and one epoch from 32 slots
+    pub fn get_light_client_update(&self, period: u64) -> Result<LightClientUpdate, Box<dyn Error>> {
+        let url = format!("{}/{}?start_period={}&count=1", self.endpoint_url, Self::GET_LIGHT_CLIENT_UPDATE_API, period);
```
Comment (sept-en) [inline eth2near/eth2near-block-relay-rs/src/beacon_rpc_client.rs:142]:
```suggestion
        let url = format!("{}/{}?start_period={}&count=1", self.endpoint_url, Self::URL_GET_LIGHT_CLIENT_UPDATE_API, period);
```

File: eth2near/eth2near-block-relay-rs/src/beacon_rpc_client.rs [hunk 5]
```diff
@@ -49,21 +116,119 @@ impl BeaconRPCClient {
...
+            signature_slot: self.get_signature_slot(&light_client_update_json_str)?,
+            finality_update: self.get_finality_update_from_light_client_update_json_str(&light_client_update_json_str)?,
+            sync_committee_update: None::<SyncCommitteeUpdate>,
+        })
+    }
+
+    pub fn get_finality_light_client_update_with_sync_commity_update(&self) -> Result<LightClientUpdate, Box<dyn Error>> {
+        let url_finality = format!("{}/{}", self.endpoint_url, Self::URL_FINALITY_LIGHT_CLIENT_UPDATE_PATH);
+        let last_period = Self::get_period_for_slot(self.get_last_slot_number()?.as_u64());
+        let url_update = format!("{}/{}?start_period={}&count=1", self.endpoint_url, Self::GET_LIGHT_CLIENT_UPDATE_API, last_period);
```
Comment (sept-en) [inline eth2near/eth2near-block-relay-rs/src/beacon_rpc_client.rs:202]:
```suggestion
        let url_update = format!("{}/{}?start_period={}&count=1", self.endpoint_url, Self::URL_GET_LIGHT_CLIENT_UPDATE_API, last_period);
```

File: eth2near/eth2near-block-relay-rs/src/beacon_rpc_client.rs [hunk 6]
```diff
@@ -49,21 +116,119 @@ impl BeaconRPCClient {
...
+            Self::get_header_json_from_rpc_result(&self.get_json_from_raw_request(&url)?)?;
         Ok(serde_json::from_str(&json_str)?)
     }
 
+    /// Returns `LightClientUpdate` struct for the given `period`.
+    ///
+    /// # Arguments
+    ///
+    /// * `period` - period id for which LightClientUpdate is fetched.
+    /// In Mainnet one period consist of 256 epochs and one epoch from 32 slots
```
Comment (sept-en) [inline eth2near/eth2near-block-relay-rs/src/beacon_rpc_client.rs:140]:
```suggestion
    /// * `period` - period id for which `LightClientUpdate` is fetched.
    /// On Mainnet, one period consists of 256 epochs, and one epoch consists of 32 slots
```

File: eth2near/eth2near-block-relay-rs/src/beacon_rpc_client.rs [hunk 7]
```diff
@@ -79,23 +244,109 @@ impl BeaconRPCClient {
...
+        let v: Value = serde_json::from_str(light_client_update_json_str)?;
+        let sync_aggregate_json_str = serde_json::to_string(&v["data"][0]["sync_aggregate"])?;
+        let sync_aggregate: SyncAggregate = serde_json::from_str(&sync_aggregate_json_str)?;
+
+        Ok(sync_aggregate)
+    }
+
+    // signature_slot doesn't provided in current API. The slot is bruteforced
+    // until SyncAggregate in BeconBlockBody in the current slot is equal
+    // to SyncAggregate in LightClientUpdate
```
Comment (sept-en) [inline eth2near/eth2near-block-relay-rs/src/beacon_rpc_client.rs:267]:
```suggestion
    // `signature_slot` is not provided in the current API. The slot is brute-forced
    // until `SyncAggregate` in `BeconBlockBody` in the current slot is equal
    // to `SyncAggregate` in `LightClientUpdate`
```

File: eth2near/eth2near-block-relay-rs/src/beacon_rpc_client.rs [hunk 8]
```diff
@@ -234,4 +477,77 @@ mod tests {
...
+        // check sync_committe_update
+        let sync_committe_update = light_client_update.sync_committee_update.unwrap();
+        assert_eq!(serde_json::to_string(&sync_committe_update.next_sync_committee.aggregate_pubkey).unwrap(), "\"0x8dcf66eb3c34854131fd7926cd86c0d10051ad045e1fbc30c59e9162019b46f2daa0e0914bf1b32541dfa54b7be57c7c\"");
+        assert_eq!(serde_json::to_string(&sync_committe_update.next_sync_committee_branch[1]).unwrap(), "\"0xedeb16e5754a4be920bb51e97dbf15833f838a5770e8509cc34cde12ee74422e\"");
+    }
+
+    //function which print json for last LightClientUpdate
+    #[test]
+    #[ignore]
+    fn show_get_light_client_update() {
```
Comment (sept-en) [inline eth2near/eth2near-block-relay-rs/src/beacon_rpc_client.rs:516]:
```suggestion
    // a utility function which prints JSON for last `LightClientUpdate`
    #[test]
    #[ignore]
    fn utility_show_get_light_client_update() {
```

File: eth2near/eth2near-block-relay-rs/src/beacon_rpc_client.rs [hunk 9]
```diff
@@ -234,4 +477,77 @@ mod tests {
...
+        let light_client_update = light_client_update_fetcher.get_light_client_update(period).unwrap();
+        let light_client_update_json_str = serde_json::to_string(&light_client_update).unwrap();
+
+        println!("Light client update pariod={}: {}", period, light_client_update_json_str);
+    }
+
+    //function which prints jsons for all BeaconBlockHeaders with Execution Data in specific range
+    #[test]
+    #[ignore]
+    fn show_headers_jsons_for_light_client_update() {
```
Comment (sept-en) [inline eth2near/eth2near-block-relay-rs/src/beacon_rpc_client.rs:529]:
```suggestion
    // a utility function that prints JSON strings for all `BeaconBlockHeader`s with `ExecutionData` in specific range
    #[test]
    #[ignore]
    fn utility_show_headers_jsons_for_light_client_update() {
```

File: eth2near/eth2near-block-relay-rs/src/config.rs [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+    // The ethereum network name (main, kiln)
+    pub network: String,
+
+    // Contract type (near, dao, file)
+    pub contract_type: String,
+
+    // Frequency of submission light client updates. Once in N epochs. 
+    pub light_client_updates_submission_frequency_in_epochs: i64,
+}
```
Comment (sept-en) [inline eth2near/eth2near-block-relay-rs/src/config.rs:36]:
```suggestion
    // endpoint to a full node of Eth2 Beacon chain with Light Client API
    pub beacon_endpoint: String,

    // endpoint for the Ethereum full node, which supports Eth1 RPC API
    pub eth1_endpoint: String,

    // the max number of headers submitted in one batch to eth client
    pub total_submit_headers: u32,

    // endpoint for a full node on the NEAR chain
    pub near_endpoint: String,

    // Account id from which relay make requests
    pub signer_account_id: String,

    // Path to the file with a secret key for signer account
    pub path_to_signer_secret_key: String,

    // Account id for eth client contract on NEAR
    pub contract_account_id: String,

    // The Ethereum network name (main, kiln)
    pub network: String,

    // Contract type (near, dao, file)
    pub contract_type: String,

    // Frequency of submission light client updates. Once in N epochs. 
    pub light_client_updates_submission_frequency_in_epochs: i64,
}
```

File: eth2near/eth2near-block-relay-rs/src/execution_block_proof.rs [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+use types::{BeaconBlockBody, MainnetEthSpec};
+
+/// `ExecutionBlockProof` contains a `block_hash` (execution block) and
+/// a proof of its inclusion in `BeaconBlockBody` tree hash.
+/// The block_hash is the 12th field in execution_payload which the 9th field in BeaconBlockBody
+/// The first 4 elements in proof correspondent to the proof of inclusion of block_hash in
+/// Merkle tree built for ExecutionPayload.
+/// The last 4 elements the proof of `ExecutionPayload` in Merkle tree is built
+/// on high-level `BeaconBlockBody` fields.
+/// The proof starts from the leaf.
```
Comment (sept-en) [inline eth2near/eth2near-block-relay-rs/src/execution_block_proof.rs:14]:
```suggestion
/// `ExecutionBlockProof` contains a `block_hash` (execution block) and
/// a proof of its inclusion in the `BeaconBlockBody` tree hash.
/// The `block_hash` is the 12th field in execution_payload, which is the 9th field in `BeaconBlockBody`
/// The first 4 elements in proof correspondent to the proof of inclusion of `block_hash` in
/// Merkle tree built for `ExecutionPayload`.
/// The last 4 elements of the proof of `ExecutionPayload` in the Merkle tree are built
/// on high-level `BeaconBlockBody` fields.
/// The proof starts from the leaf.
```

File: eth2near/finality-update-verify/Cargo.toml [hunk 12]
```diff
@@ -0,0 +1,23 @@
+[package]
+name = "finality-update-verify"
+version = "0.1.0"
+edition = "2021"
+
+[dependencies]
+eth-types = { path ="../../contracts/near/eth-types/", features = ["eip1559"]}
+bls = { git = "https://github.com/aurora-is-near/lighthouse.git", branch = "wasm-tmp", default-features = false, features = ["milagro"]}
+eth2-client  = { path ="../../contracts/near/eth2-client"}
```
Comment (sept-en) [inline eth2near/finality-update-verify/Cargo.toml:9]: It seems, as only `utility` mod is used from `eth2-client`, we need to use `utility` crate explicitly.

**Review by sept-en — CHANGES_REQUESTED** [review]

File: eth2near/contract_wrapper/src/eth_client_contract.rs [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+            network,
+            finalized_execution_header,
+            finalized_beacon_header,
+            current_sync_committee,
+            next_sync_committee,
+            validate_updates: true,
+            verify_bls_signatures: false,
+            hashes_gc_threshold: 51000,
+            max_submitted_blocks_by_account: 8000,
+            trusted_signer: Option::<AccountId>::None,
```
[the remark this answers is not in the document]
Comment (sept-en) [reply eth2near/contract_wrapper/src/eth_client_contract.rs:103]: Created an issue https://github.com/aurora-is-near/rainbow-bridge/issues/769 for that

File: eth2near/eth2near-block-relay-rs/olga_local_test_config.toml (hunk 2 above)
Comment (sept-en) [reply eth2near/eth2near-block-relay-rs/olga_local_test_config.toml]: Fixed in 40e8b0f439def64dff0c44ac0c18f6f3172f75f7 and 1ed11716c2ebb1227c7d8056398d97520ab5bbc7

File: eth2near/finality-update-verify/Cargo.toml (hunk 12 above)
Comment (sept-en) [reply eth2near/finality-update-verify/Cargo.toml]: Merged `the-merge` branch and fixed in 7a4de8829fb6e854a2a9f41165aa6a0eddb3bba1

File: eth2near/contract_wrapper/src/near_contract_wrapper.rs [hunk 14]
```diff
@@ -81,13 +81,15 @@ impl ContractWrapper for NearContractWrapper {
             _ => Err("failed to extract current nonce")?,
         };
 
+        let num_blocks_in_batch = method_name.len() as u64;
+        let attached_gas_per_promise_in_batch = 300_000_000_000_000 / num_blocks_in_batch;
```
Comment (karim-en) [inline eth2near/contract_wrapper/src/near_contract_wrapper.rs:85]: I would recommend using `parse_gas!("300 T") ` from https://docs.rs/near-units/latest/near_units/

**Review by karim-en — APPROVED** [review]

File: eth2near/contract_wrapper/src/near_contract_wrapper.rs (hunk 14 above)
Comment (sept-en) [reply eth2near/contract_wrapper/src/near_contract_wrapper.rs:85]: Created issue https://github.com/aurora-is-near/rainbow-bridge/issues/770

**Review by sept-en — APPROVED** [review]

**Pull request merged**